code/go/internal/validator/folder_item_spec.go

```diff
@@ -0,0 +1,166 @@
+package validator
+
+import (
+	"encoding/json"
+	"fmt"
+	"gopkg.in/yaml.v3"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path/filepath"
+	"regexp"
+
+	"github.com/pkg/errors"
+	"github.com/xeipuuv/gojsonschema"
+)
+
+type folderItemSpec struct {
+	Description      string `yaml:"description"`
+	ItemType         string `yaml:"type"`
+	ContentMediaType string `yaml:"contentMediaType"`
+	Name             string `yaml:"name"`
+	Pattern          string `yaml:"pattern"`
+	Required         bool   `yaml:"required"`
+	Ref              string `yaml:"$ref"`
+	Visibility       string `yaml:"visibility" default:"public"`
+	commonSpec       `yaml:",inline"`
+}
+
+type itemSchemaSpec struct {
+	Spec map[string]interface{} `json:"spec" yaml:"spec"`
+}
+
+func (s *folderItemSpec) matchingFileExists(files []os.FileInfo) (bool, error) {
+	if s.Name != "" {
+		for _, file := range files {
+			if file.Name() == s.Name {
+				return s.isSameType(file), nil
+			}
+		}
+	} else if s.Pattern != "" {
+		for _, file := range files {
+			isMatch, err := regexp.MatchString(s.Pattern, file.Name())
+			if err != nil {
+				return false, errors.Wrap(err, "invalid folder item spec pattern")
+			}
+			if isMatch {
+				return s.isSameType(file), nil
+			}
+		}
+	}
+
+	return false, nil
+}
+
+func (s *folderItemSpec) isSameType(file os.FileInfo) bool {
+	switch s.ItemType {
+	case itemTypeFile:
+		return !file.IsDir()
+	case itemTypeFolder:
+		return file.IsDir()
+	}
+
+	return false
+}
+
+func (s *folderItemSpec) validate(fs http.FileSystem, folderSpecPath string, itemPath string) ValidationErrors {
+	if s.Ref == "" {
+		return nil // no item's schema defined
+	}
+
+	schemaPath := filepath.Join(filepath.Dir(folderSpecPath), s.Ref)
+	schemaData, err := loadItemSchema(fs, schemaPath)
+	if err != nil {
+		return ValidationErrors{errors.Wrapf(err, "loading item schema failed (path %s)", schemaPath)}
+	}
+
+	// loading item content
+	itemData, err := loadItemContent(itemPath, s.ContentMediaType)
+	if err != nil {
+		return ValidationErrors{errors.Wrapf(err, "loading item content failed (path %s)", itemPath)}
+	}
+
+	// validation with schema
+	errs := validateData(schemaData, itemData)
+	if errs != nil {
+		return errs
+	}
+	return nil
+}
+
+func loadItemSchema(fs http.FileSystem, itemSchemaPath string) ([]byte, error) {
+	itemSchemaFile, err := fs.Open(itemSchemaPath)
+	if err != nil {
+		return nil, errors.Wrap(err, "opening schema file failed")
+	}
+	defer itemSchemaFile.Close()
+
+	itemSchemaData, err := ioutil.ReadAll(itemSchemaFile)
```

Comment:
Any reason not to use `ioutil.ReadFile` here directly instead of opening the file handle with `fs.Open` first and then reading from it using `ioutil.ReadAll`?

Reply:
We need to read data from statik resource. I'm afraid that `ioutil.ReadFile` doesn't support it this way.

Reply:
Ha, I got tripped up by the `fs` there! Totally makes sense why you have it the way it is. Ignore me. 🙂